src/libsyntax/feature_gate.rs

```diff
@@ -718,6 +719,41 @@ pub enum AttributeGate {
     Ungated,
 }
 
+#[derive(Clone, Copy)]
+pub struct AttributeTemplate {
+    word: bool,
+    list: Option<&'static str>,
+    name_value_str: Option<&'static str>,
+}
+
+impl AttributeTemplate {
+    fn compatible(&self, meta_item_kind: &ast::MetaItemKind) -> bool {
+        match meta_item_kind {
+            ast::MetaItemKind::Word => self.word,
+            ast::MetaItemKind::List(..) => self.list.is_some(),
+            ast::MetaItemKind::NameValue(lit) if lit.node.is_str() => self.name_value_str.is_some(),
+            ast::MetaItemKind::NameValue(..) => false,
+        }
+    }
+}
+
+macro_rules! template {
+    (@ $word: expr, $list: expr, $name_value_str: expr) => { AttributeTemplate {
```

Comment:
I would move this arm to the end, since it's the one you are least likely to want to read when skimming through